### PR TITLE
[stdlib] now `strip` can take `chars` argument and `String` now have useful aliases

### PR DESCRIPTION
**Pull request by ormeilu**
Two major changes:
1. `strip`, `lstrip` and `rstrip` now can remove custom chars
2. `String` now have useful aliases like Python's https://github.com/python/cpython/blob/0085c3ae8f067abd4f6540d0f6dd2fb13107618e/Lib/string.py#L24C1-L32C62

- Naming is identical to python's
- Defaults to whitespace ` \n\t\r\f\v` without casting to Int8
- Tests are updated

### CODE REVIEW

File: stdlib/src/builtin/string.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/builtin/string.mojo:1211]: **suggestion** since the list of characters for whitespaces is fixed, we could define somewhere `alias WHITESPACES = " \n\t\r\f\v"` and use it in the default values to ensure the behavior is the same.

File: stdlib/src/builtin/string.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/builtin/string.mojo:1250]: **thoughts** In the default use case, we might end up with a slower function. Before, it was checking that the byte was in a given interval (something like `return a < x < b`, see the implementation of `isspace()`) and now it's checking the value against one of the 5 elements in the list. With simd, the performance might be similar. 

I don't believe it needs to be changed right now with overloads for more performance, I'm just mentionning for the maintainers that we can improve in two areas:
* Ensure the same code path is taken for isspace and the default use of `rstrip`, `lstrip`, to ensure we don't have different behaviors
* Have a benchmarking tool to allow us to verify that we don't have regressions. Here it's not clear if we'll get a performance regression or not. (it might not be the focus this early so it might not be too important).

**Review by gabrieldemarmiesse — COMMENTED** [review]
Many thanks for the PR, this is a nice addition! The tests look good :) Let me know if you have trouble with the DCO

File: stdlib/src/builtin/string.mojo
Comment (toiletsandpaper) [reply stdlib/src/builtin/string.mojo:1211]: Yeah, good idea. I thought about it, but forgot somehow. I'll probably add everything from https://github.com/python/cpython/blob/0085c3ae8f067abd4f6540d0f6dd2fb13107618e/Lib/string.py#L24C1-L32C62 now and will add additional tests

Comment (toiletsandpaper) [reply stdlib/src/builtin/string.mojo]: Done.

Comment (gabrieldemarmiesse) [reply stdlib/src/builtin/string.mojo]: Maybe you can add other the constants in a separate pull request? I believe it will be more controversial than just `whitespaces`. Especially since there are functions already in the codebase that check for lowercase, uppercase, etc and those functions do not currently use those constants, they use a more optimized algorithm.

We wouldn't want to block the merge of strip, rstrip and lstrip because of a discussion concerning those new constants.

File: stdlib/src/builtin/string.mojo
Comment (toiletsandpaper) [reply stdlib/src/builtin/string.mojo:1261]: Yep, that will definitely slow down checks a bit, but not so much. But for now it's the only readable and flexible solution at the same time I see.

**Comment by gabrieldemarmiesse** [comment]
You might want to sign-off your commits or the DCO CI job will not pass. See https://github.com/modularml/mojo/pull/2555/checks?check_run_id=24632558478 for the instructions.

File: stdlib/src/builtin/string.mojo
Comment (gabrieldemarmiesse) [reply stdlib/src/builtin/string.mojo:1261]: Indeed I'm not asking for changes, it's just a note for the maintainers

File: stdlib/src/builtin/string.mojo
Comment (toiletsandpaper) [reply stdlib/src/builtin/string.mojo]: I don't really know how to open different PR from same branch on GitHub.

About functions - python already have functions and constants too

Comment (toiletsandpaper) [reply stdlib/src/builtin/string.mojo]: Btw, current implementation is using `String.WHITESPACE`

**Comment by toiletsandpaper** [comment]
> You might want to sign-off your commits or the DCO CI job will not pass. See https://github.com/modularml/mojo/pull/2555/checks?check_run_id=24632558478 for the instructions.

Thx. I thought GPG auto-sign would be enough

File: stdlib/src/builtin/string.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/builtin/string.mojo:450]: A note for the maintainers: I don't have an opinion here, but I'm just pointing out that in Python, those constants are lowercase: https://docs.python.org/3/library/string.html#string-constants . We just need to be aware of this fact before merging.

File: stdlib/src/builtin/string.mojo
Comment (gryznar) [inline stdlib/src/builtin/string.mojo:364]: What is the purpose to break consistency and not to keep snake_case? This is internal functionality. There is no need to keep this aligned with Python. Same to below

File: stdlib/src/builtin/string.mojo
Comment (gabrieldemarmiesse) [inline stdlib/src/builtin/string.mojo:1470]: I believe that would break the URL here, the "search and replace" might have been a bit too quick 😄

File: stdlib/src/builtin/string.mojo
Comment (gryznar) [reply stdlib/src/builtin/string.mojo:450]: I would prefer to keep them uppercase. But this is question to Joe Loser I think

File: stdlib/src/builtin/string.mojo
Comment (gryznar) [inline stdlib/src/builtin/string.mojo:447]: If we would not follow Python, this can be renamed to: HEX_DIGITS and OCTDIGITS -> OCT_DIGITS

File: stdlib/src/builtin/string.mojo
Comment (gabrieldemarmiesse) [reply stdlib/src/builtin/string.mojo]: You could make two branches if you want. https://github.com/modularml/mojo/pull/2555#discussion_r1591073450 , https://github.com/modularml/mojo/pull/2555#discussion_r1591067735 show that those new constants are more controversial than the strip function improvements. It could have value to separate those two changes (one PR for String.whitespace + strip,rstrip,lstrip and one PR for all the new constants) otherwise everything will be blocked until all discussions have been resolved. (it's just a tip from an old OSS dude to get your PRs merged faster, you're not obligated to follow it at all)

File: stdlib/src/builtin/string.mojo
Comment (toiletsandpaper) [reply stdlib/src/builtin/string.mojo]: That was accidental by using `change all occurences` instead of normal renaming in my IDE.
Good catch, thx!

File: stdlib/src/builtin/string.mojo
Comment (toiletsandpaper) [reply stdlib/src/builtin/string.mojo]: That was accidental by using `change all occurences` instead of normal renaming in my IDE.
Good catch, thx!

File: stdlib/src/builtin/string.mojo
Comment (toiletsandpaper) [reply stdlib/src/builtin/string.mojo]: That was accidental by using `change all occurences` instead of normal renaming in my IDE.
Good catch, thx!

File: stdlib/src/builtin/string.mojo
Comment (toiletsandpaper) [reply stdlib/src/builtin/string.mojo]: Yep, looks good to me, because we're already breaking python naming here, we can change this as well.

**Comment by toiletsandpaper** [comment]
Am I missing something that needs to be done or it's just "sit and wait" part?

File: stdlib/test/builtin/test_string.mojo
Comment (JoeLoser) [inline stdlib/test/builtin/test_string.mojo:734]: **Suggestion** Remove this `print`.  It's not needed for anything.

File: stdlib/test/builtin/test_string.mojo
Comment (JoeLoser) [inline stdlib/test/builtin/test_string.mojo:678]: **Suggestion** This is good, but do you mind also adding a negative test (i.e. testing for a character that is not printable)?

**Review by JoeLoser — APPROVED** [review]
This is a great addition, thank you!

File: stdlib/test/builtin/test_string.mojo
Comment (toiletsandpaper) [reply stdlib/test/builtin/test_string.mojo]: Done :)

File: stdlib/test/builtin/test_string.mojo
Comment (toiletsandpaper) [reply stdlib/test/builtin/test_string.mojo:678]: Done :)

Comment (JoeLoser) [reply stdlib/test/builtin/test_string.mojo:678]: Thanks! I just re-imported this internally and it should land here in a little bit.

**Comment by JoeLoser** [comment]
Merged today in https://github.com/modularml/mojo/commit/682f225b164c661ea79ccc056346d662ef3e5d54 - thanks for the contribution!

**Pull request closed without merging**
